File: src/domain/usecases/folders/UpdateFolderUseCase.ts
import type { ImageFolderRepository } from '@/domain/repositories/ImageFolderRepository'
import type { UserRepository } from '@/domain/repositories/UserRepository'
import type { ImageFolder } from '@/domain/entities/ImageFolder'
import type { User } from '@/domain/entities/User'
import { assertUserCanEditFolder } from '@/domain/entities/User'
import {
  NotFoundError,
  UnauthorizedError,
  ValidationError,
} from '@/domain/errors/DomainError'
import { resolveAssignments } from '@/domain/usecases/folders/CreateFolderUseCase'

export interface UpdateFolderRequest {
  folderId: string
  name: string
  description: string
  assignToAllTechnicians: boolean
  assignedTechnicianIds: string[]
}

export class UpdateFolderUseCase {
  private readonly folderRepository: ImageFolderRepository
  private readonly userRepository: UserRepository

  constructor(
    folderRepository: ImageFolderRepository,
    userRepository: UserRepository,
  ) {
    this.folderRepository = folderRepository
    this.userRepository = userRepository
  }

  async execute(actor: User, request: UpdateFolderRequest): Promise<ImageFolder> {
    const folder = await this.folderRepository.getById(request.folderId)
    if (!folder) {
      throw new NotFoundError('Carpeta no encontrada')
    }

    if (!assertUserCanEditFolder(actor, folder)) {
      throw new UnauthorizedError('No tienes permiso para editar esta carpeta')
    }

    const name = folder.routeCode
      ? folder.name
      : request.name.trim()
    const description = request.description.trim()

    if (!name) {
      throw new ValidationError('El nombre de la carpeta es obligatorio')
    }

    const assignment = await resolveAssignments(
      this.userRepository,
      actor,
      request.assignToAllTechnicians,
      request.assignedTechnicianIds,
    )

    return this.folderRepository.update(request.folderId, {
      name,
      description,
      assignToAllTechnicians: assignment.assignToAllTechnicians,
      assignedTechnicianIds: assignment.assignedTechnicianIds,
      assignedTechnicianNames: assignment.assignedTechnicianNames,
    })
  }
}

File: consorcio/lib/domain/usecases/update_folder_use_case.dart
import '../entities/app_user.dart';
import '../entities/image_folder.dart';
import '../errors/domain_exception.dart';
import '../repositories/image_folder_repository.dart';
import '../repositories/user_repository.dart';
import 'create_folder_use_case.dart';

class UpdateFolderUseCase {
  UpdateFolderUseCase(this._folderRepository, this._userRepository);
  final ImageFolderRepository _folderRepository;
  final UserRepository _userRepository;

  Future<ImageFolder> execute(
    AppUser actor, {
    required String folderId,
    required String name,
    required String description,
    required bool assignToAllTechnicians,
    required List<String> assignedTechnicianIds,
  }) async {
    actor.assertCanOperateApp();

    final folder = await _folderRepository.getById(folderId);
    if (folder == null) {
      throw DomainException('Carpeta no encontrada');
    }
    if (!folder.canBeAccessedBy(actor.id)) {
      throw DomainException('No puedes editar esta carpeta');
    }

    final cleanName = name.trim();
    if (cleanName.isEmpty) {
      throw DomainException('El nombre de la carpeta es obligatorio');
    }

    final assignment = await resolveFolderAssignments(
      userRepository: _userRepository,
      actor: actor,
      assignToAllTechnicians: assignToAllTechnicians,
      assignedTechnicianIds: assignedTechnicianIds,
    );

    return _folderRepository.update(
      id: folderId,
      name: cleanName,
      description: description.trim(),
      assignToAllTechnicians: assignment.assignToAll,
      assignedTechnicianIds: assignment.ids,
      assignedTechnicianNames: assignment.names,
    );
  }
}
